Add support for InputDeviceInfo

youennf opened the pull request:
#### e89f5a6ca7978ba955a84e74f7afdcd8c4668a33
<pre>
Add support for InputDeviceInfo
<a href="https://bugs.webkit.org/show_bug.cgi?id=253896">https://bugs.webkit.org/show_bug.cgi?id=253896</a>
rdar://problem/106716244

Reviewed by Eric Carlson.

Expose InputDeviceInfo to help web applications sleect the right device and configuration.
To do this, UIProcess is now computing capabilities by creating a source when enumerating each device.
It is sent through IPC to WebProcess that will create the InputDeviceInfo object.
We update IDL to expose InputDeviceInfo.

Covered by WPT tests.

* LayoutTests/fast/mediastream/media-device-info.html:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/MediaDevices-enumerateDevices.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/MediaStreamTrack-getCapabilities.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/idlharness.https.window-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/InputDeviceInfo.cpp: Added.
(WebCore::InputDeviceInfo::InputDeviceInfo):
(WebCore::InputDeviceInfo::getCapabilities const):
* Source/WebCore/Modules/mediastream/InputDeviceInfo.h: Added.
* Source/WebCore/Modules/mediastream/InputDeviceInfo.idl: Added.
* Source/WebCore/Modules/mediastream/MediaDeviceInfo.cpp:
(WebCore::toMediaDeviceInfoKind):
* Source/WebCore/Modules/mediastream/MediaDeviceInfo.h:
(): Deleted.
* Source/WebCore/Modules/mediastream/MediaDeviceInfo.idl:
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::MediaDevices::exposeDevices):
(WebCore::MediaDevices::enumerateDevices):
(WebCore::toMediaDeviceInfoKind): Deleted.
* Source/WebCore/Modules/mediastream/MediaDevices.h:
* Source/WebCore/Modules/mediastream/MediaDevices.idl:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getCapabilities const):
(WebCore::capabilityDoubleRange): Deleted.
(WebCore::capabilityIntRange): Deleted.
(WebCore::capabilityStringVector): Deleted.
(WebCore::capabilityBooleanVector): Deleted.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp: Added.
(WebCore::capabilityDoubleRange):
(WebCore::capabilityIntRange):
(WebCore::capabilityStringVector):
(WebCore::capabilityBooleanVector):
(WebCore::toMediaTrackCapabilities):
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.h: Added.
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.idl: Added.
* Source/WebCore/Modules/mediastream/UserMediaClient.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/platform/mediastream/CaptureDeviceWithCapabilities.h: Added.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::getCapabilities):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::haveMicrophoneDevice):
(WebKit::UserMediaPermissionRequestManagerProxy::platformGetMediaStreamDevices):
(WebKit::UserMediaPermissionRequestManagerProxy::computeFilteredDeviceList):
(WebKit::UserMediaPermissionRequestManagerProxy::enumerateMediaDevicesForFrame):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::enumerateMediaDevicesForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/glib/UserMediaPermissionRequestManagerProxyGLib.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::platformGetMediaStreamDevices):
* Source/WebKit/WebProcess/MediaStream/UserMediaPermissionRequestManager.cpp:
(WebKit::UserMediaPermissionRequestManager::enumerateMediaDevices):
* Source/WebKit/WebProcess/MediaStream/UserMediaPermissionRequestManager.h:

Canonical link: <a href="https://commits.webkit.org/261745@main">https://commits.webkit.org/261745@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/82a05204401b0f218753411201b6cbafd362376a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/112706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21873 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1388 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4493 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121243 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/116776 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13038 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5656 "Passed tests") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/118475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100481 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105796 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99194 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/1025 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46261 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14192 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1065 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14880 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20210 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/53057 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8192 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16725 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/Modules/mediastream/InputDeviceInfo.cpp:
```diff
@@ -0,0 +1,51 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(InputDeviceInfo);
+
+InputDeviceInfo::InputDeviceInfo(CaptureDeviceWithCapabilities&& deviceWithCapabilities, String&& deviceId, String&& groupId)
```
`deviceWithCapabilities.device` has a deviceId and a groupId. Can yo use those, or are `deviceId` and `groupId` the salted IDs? If so, it might be worth naming them to make that clear.

youennf replied:
I changed the name

eric-carlson commented on Source/WebCore/Modules/mediastream/MediaDeviceInfo.cpp:
```diff
@@ -47,6 +45,25 @@ MediaDeviceInfo::MediaDeviceInfo(const String& label, const String& deviceId, co
+        ASSERT_NOT_REACHED();
+        return MediaDeviceInfo::Kind::Audioinput;
+    }
+    return MediaDeviceInfo::Kind::Audioinput;
```
Nit: you could `break` in the unknown case and assert before returning.

youennf replied:
Done

eric-carlson commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -3,7 +3,7 @@
 	classes = {
 	};
-	objectVersion = 51;
+	objectVersion = 54;
```
You may not want to change this.

youennf replied:
Done

eric-carlson reviewed (approved):
r=me

youennf committed:
Add support for InputDeviceInfo
https://bugs.webkit.org/show_bug.cgi?id=253896
rdar://problem/106716244

Reviewed by Eric Carlson.

Expose InputDeviceInfo to help web applications sleect the right device and configuration.
To do this, UIProcess is now computing capabilities by creating a source when enumerating each device.
It is sent through IPC to WebProcess that will create the InputDeviceInfo object.
We update IDL to expose InputDeviceInfo.

Covered by WPT tests.

* LayoutTests/fast/mediastream/media-device-info.html:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/MediaDevices-enumerateDevices.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/MediaStreamTrack-getCapabilities.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-streams/idlharness.https.window-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/InputDeviceInfo.cpp: Added.
(WebCore::InputDeviceInfo::InputDeviceInfo):
(WebCore::InputDeviceInfo::getCapabilities const):
* Source/WebCore/Modules/mediastream/InputDeviceInfo.h: Added.
* Source/WebCore/Modules/mediastream/InputDeviceInfo.idl: Added.
* Source/WebCore/Modules/mediastream/MediaDeviceInfo.cpp:
(WebCore::toMediaDeviceInfoKind):
* Source/WebCore/Modules/mediastream/MediaDeviceInfo.h:
(): Deleted.
* Source/WebCore/Modules/mediastream/MediaDeviceInfo.idl:
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::MediaDevices::exposeDevices):
(WebCore::MediaDevices::enumerateDevices):
(WebCore::toMediaDeviceInfoKind): Deleted.
* Source/WebCore/Modules/mediastream/MediaDevices.h:
* Source/WebCore/Modules/mediastream/MediaDevices.idl:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getCapabilities const):
(WebCore::capabilityDoubleRange): Deleted.
(WebCore::capabilityIntRange): Deleted.
(WebCore::capabilityStringVector): Deleted.
(WebCore::capabilityBooleanVector): Deleted.
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp: Added.
(WebCore::capabilityDoubleRange):
(WebCore::capabilityIntRange):
(WebCore::capabilityStringVector):
(WebCore::capabilityBooleanVector):
(WebCore::toMediaTrackCapabilities):
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.h: Added.
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.idl: Added.
* Source/WebCore/Modules/mediastream/UserMediaClient.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/platform/mediastream/CaptureDeviceWithCapabilities.h: Added.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::getCapabilities):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::haveMicrophoneDevice):
(WebKit::UserMediaPermissionRequestManagerProxy::platformGetMediaStreamDevices):
(WebKit::UserMediaPermissionRequestManagerProxy::computeFilteredDeviceList):
(WebKit::UserMediaPermissionRequestManagerProxy::enumerateMediaDevicesForFrame):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::enumerateMediaDevicesForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/glib/UserMediaPermissionRequestManagerProxyGLib.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::platformGetMediaStreamDevices):
* Source/WebKit/WebProcess/MediaStream/UserMediaPermissionRequestManager.cpp:
(WebKit::UserMediaPermissionRequestManager::enumerateMediaDevices):
* Source/WebKit/WebProcess/MediaStream/UserMediaPermissionRequestManager.h:

Canonical link: https://commits.webkit.org/261745@main
